### ISSUE TITLE
Forms with dynamic content impossible in featherlight?

### ISSUE THREAD

**Issue by rogerlos** [noelboss/featherlight#122]
As far as I can tell, when you open existing (but hidden) DOM content in a featherlight modal, you completely duplicate the target DOM elements--without removing or renaming those elements? 

Form fields which have to be referred to by ID (for example, to attach an editor to) cannot be updated in the featherlight window, as duplicate IDs exist in the DOM at that point.

I'm probably just missing something crucial here.

**Comment by xStrom** [comment]
The DOM content duplication issue is real and has been causing quite a few problems for me as well.

One way to solve it is to write relative javascript, not using ids. Say you have a button which needs to take the value of an input and then display it. Instead of using the input id, traverse the DOM from the button itself. Something like this:

    $(this).prev("input").val()

I do wonder however if it's already possible or how easy it would be to add a feature to featherlight where you define your DOM content in a specific area that featherlight knows how to handle, thus removing the need for any duplication.

**Comment by xStrom** [comment]
Another quick thought regarding removing duplication.

Perhaps there's an option, or one could be added, which when set would move the specified content instead of copying it. Then there will only be a single copy.

Future hide/show could be done exactly that way - by hiding/showing the already moved content.

**Comment by marcandre** [comment]
It's true that the ID thing can be problematic.

Featherlight could move instead of copy, and move back when done

**Comment by marcandre** [comment]
I also have a need to reopen a featherlight with the previous content (valid for all contentFilters though, not just `jquery`), so now I'm wondering what's the best API for this.

**Comment by rogerlos** [comment]
I ended up removing the hidden DOM elements once featherlight opens, and instead of having them in a static hidden div, using a javascript template engine to generate them on the fly. Seems like a number of extra steps, and, of course, precludes having an easily accessed "show me the last version of this" copy hanging around.

The relative idea has merit, but in my case I need to attach tinymce editors to some textareas, and that requires IDs. If you don't have to use IDs, you can use the `.featherlight-content` class as a selector to grab featherlight version.

Moving the content would be ideal, of course.

**Comment by marcandre** [comment]
Ok, so here's the two behaviors I'm thinking about:
1) (current)
  -jquery content filter uses `clone(true)`
  -content refreshed each time the dialog is opened, `getContent` & `setContent` called each time, with the callbacks `beforeContent`, and `afterContent`.
2) (new option)
  - jquery content filter moves instead of clone
  - content is kept intact. If the dialog is reopened, the same content is displayed. `getContent` & `setContent` are not called again. `beforeContent` and `afterContent` should probably not be called again, only `beforeOpen` and `afterOpen` are called each time. Probably no need for an additional callback like `refreshContent`.

When binding many items with that new option, I imagine the content would be unique to each item (instead of shared for all items). But `$('many items').featherlight('#someId', {newOption: true})` would not work...

The big question is how do we name the new option... Ideas?

**Comment by xStrom** [comment]
I think a good option name would be something that gives a good idea of what it does without looking at the docs. Perhaps something like ```moveAndPersist```?

**Comment by rogerlos** [comment]
I think the featherlight content reflecting ongoing modifications is the expected behavior, and that subsequent changes to the target DOM would be my own problem to deal with (for example, clearing a form before presenting it again).

So maybe the option should just be `cloneContent` or `cloneTarget` with the default being "true" to reflect the current behavior? 

That would make it fairly clear (to me, anyway) that featherlight clones content by default. The documentation can note, if the option is set to `false`, that the content-setting functions are only called the first time the modal is invoked.

**Comment by marcandre** [comment]
Thanks for the comments so far. I like the `persist` name actually. The problem I see with `clone` is that it applies only for the `jquery` content filter. But think of the `ajax` content filter, it doesn't clone anything but there's potentially a big difference between doing the ajax call on first open vs at each open, or the `iframe` filter generating and loading the `<iframe>` once vs for each trigger...

That's why the idea of `persist`ing is really what's going on here. If you don't persist, then it makes sense to `clone`, otherwise the second click wouldn't work :-) So `move` /  `clone` are just side effects of `persist` for one particular content filter.

I'm still not 100% if binding on a selection should `persist` for each item or all of them as a group. Let me see when implementing.

**Comment by marcandre** [comment]
Ok, it took only a half a dozen LOC to get it going.

I've used the name `persist` for the option, and `shared` is a special value if more than one button should share the same content. I have a feeling that usually, individual persist will be more common than shared.

Anyways, if there's a bunch of `<a data-featherlight data-featherlight-persist>` in a page, I think it makes sense they don't share their content...

What do you guys thing of this API and the following doc?

persist - Boolean or 'shared': false;
If set, the content persist and will be shown again when opened again.
In case where multiple buttons need to persist the same content, use the special value 'shared'.
The content filter `jquery` (used for links like `.some-class` or `#some-id`) will clone the given content if and only if it is not persisted. Otherwise it will be moved in the lightbox.

**Comment by xStrom** [comment]
The API looks just fine. The doc is also mostly fine, however I guess I can point out some minor issues.

Your editor's tab seems to be configured at 2 spaces so it looks fine for you, but in other cases the new persist option's comments in ```README.md``` and ```src/featherlight.js``` look misaligned. All the previous comments have been aligned by spaces, so I think the persist option's comments should be aligned using spaces as well.

Some minor grammar corrections:

    README.md

    L102 & L196
    -If set, the content persist and will
    +If set, the content will persist and will

    L198
    -it will be moved in the lightbox
    +it will be moved into the lightbox


    src/featherlight.js

    L112
    -If set, the content persist and will
    +If set, the content will persist and will

I took the persist branch out for a test run. Now you never said that the functionality is ready, but I'll mention this just in case anyway. With the persist option set, the content got moved instead of copied, however it still got deleted upon closing the lightbox and thus could no longer be opened again.

**Comment by marcandre** [comment]
Awesome, thanks for the corrections

> With the persist option set, the content got moved instead of copied, however it still got deleted upon closing the lightbox and thus could no longer be opened again.

What did you try, exactly? If you call `$.featherlight` directly, you need to keep the result in a variable and call `open()` on it the other times.

**Comment by xStrom** [comment]
I do call ```$.featherlight``` manually yes. Saving the return value and calling ```open``` on it later does work.

However the content is still missing from the DOM while the lightbox is closed. Some lighter content modifying workloads could easily be moved into the ```beforeOpen``` event handler, however I also have some pretty heavy workloads that were keeping the hidden content up-to-date and invoking featherlight would just reveal the already calculated content.

I did notice the ```$content``` property. Would using/modifying that be safe & provide the same results?

```javascript
    // Old update code, while lightbox is closed
    $("#foo p").text("bar");

    // New update code, is it equivalent?
    // Elsewhere:
    app.foo = $.featherlight($("#foo"), {persist: true});
    // While lightbox is closed
    app.foo.$content.find("p").text("bar");
```

**Comment by xStrom** [comment]
Is there something that is blocking this branch from being merged to the master and if so, can I do anything to help move it along?

(ping @marcandre)

**Comment by marcandre** [comment]
Sorry, forgot about this.

So, IIC, by "however it still got deleted upon closing the lightbox", you meant it doesn't get moved back, but it's not deleted.

Yes, you can access it with  `app.foo.$content`, or `this.$content` in callbacks.

So that seems like a good solution, right?

**Comment by xStrom** [comment]
Seems so. I'll try rewriting my usage with this branch on Monday and then I can let you know if I ran into any trouble.

**Comment by xStrom** [comment]
I have now rewritten my code to use this ```persist``` option. Thus far things seem to be working nicely and I saw a nice 40% performance gain in my specific usage as well, as measured by the Chrome profiler.

**Comment by pbowyer** [comment]
:+1: from me for merging. 

I put a WordPress Ninja Form (submitting via AJAX) in a Featherlight modal and it broke completely because of the cloning.

**Comment by marcandre** [comment]
This has been released as v1.3.0 :fireworks: 
Thanks for your patience, input and review everyone :smiley_cat:

**Comment by mateostabio** [comment]
If anyone is googling why their Hubspot Form embed isn't loading in the Featherlight JS modal popup, it's because you don't have data-featherlight-persist setting. Thanks, this fixed it for me.
